File: LowRes NX Coder/LowRes NX Coder/Scenes/Main/TabBarController.m
//
//  TabBarController.m
//  Pixels
//
//  Created by Timo Kloss on 1/10/15.
//  Copyright © 2015 Inutilis Software. All rights reserved.
//

#import "TabBarController.h"
#import "HelpSplitViewController.h"
#import "LowRes_NX_Coder-Swift.h"

@interface TabBarController ()

@end

@implementation TabBarController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    AppController.shared.tabBarController = self;
    
    UIViewController *explorerVC = [self.storyboard instantiateViewControllerWithIdentifier:@"ExplorerNav"];

    UIStoryboard *helpStoryboard = [UIStoryboard storyboardWithName:@"Help" bundle:nil];
    UIViewController *helpVC = (UIViewController *)[helpStoryboard instantiateInitialViewController];
    
    UIViewController *aboutVC = [self.storyboard instantiateViewControllerWithIdentifier:@"AboutNav"];
    
    explorerVC.tabBarItem = [self itemWithTitle:@"My Programs" imageName:@"programs"];
    helpVC.tabBarItem = [self itemWithTitle:@"Help" imageName:@"help"];
    aboutVC.tabBarItem = [self itemWithTitle:@"About" imageName:@"about"];
    
    self.viewControllers = @[explorerVC, helpVC, aboutVC];
    
    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(didAddProgram) name:@"ProjectManagerDidAddProgram" object:nil];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [AppController.shared checkShowProgram];
}

- (UITabBarItem *)itemWithTitle:(NSString *)title imageName:(NSString *)imageName
{
    return [[UITabBarItem alloc] initWithTitle:title image:[UIImage imageNamed:imageName] selectedImage:nil];
}

- (void)dismissPresentedViewController:(void (^)(void))completion
{
    UIViewController *topVC = self.selectedViewController;
    if ([topVC isKindOfClass:[UINavigationController class]])
    {
        topVC = ((UINavigationController *)topVC).topViewController;
    }
    if (topVC.presentedViewController)
    {
        [topVC dismissViewControllerAnimated:YES completion:completion];
    }
    else
    {
        completion();
    }
}

- (void)showExplorerAnimated:(BOOL)animated root:(BOOL)root
{
    self.selectedIndex = TabIndexExplorer;
    UINavigationController *nav = (UINavigationController *)self.selectedViewController;
    if (root)
    {
        [nav popToRootViewControllerAnimated:animated];
    }
    else
    {/*
        if (![nav.topViewController isKindOfClass:[ExplorerViewController class]])
        {
            [nav popViewControllerAnimated:animated];
        }*/
    }
}

- (void)showHelpForChapter:(NSString *)chapter
{
    self.selectedIndex = TabIndexHelp;
    HelpSplitViewController *helpVC = (HelpSplitViewController *)self.selectedViewController;
    [helpVC showChapter:chapter];
}

- (void)didAddProgram
{
    if (self.selectedIndex != 0)
    {
        self.selectedIndex = 0;
    }
}

@end

File: LowRes NX Coder/LowRes NX Coder/Scenes/Main/TabBarController.swift
//
//  TabBarController.swift
//  LowRes NX Coder
//
//  Created by Timo Kloss on 25/01/2020.
//  Copyright © 2020 Inutilis Software. All rights reserved.
//

import UIKit

enum TabIndex: Int {
    case explorer
    case help
    case about
    case community
}

class TabBarController: UITabBarController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        AppController.shared.tabBarController = self
        
        let explorerVC = self.storyboard!.instantiateViewController(withIdentifier: "ExplorerNav")

        let helpStoryboard = UIStoryboard(name: "Help", bundle: nil)
        let helpVC = helpStoryboard.instantiateInitialViewController()!
        
        let aboutVC = self.storyboard!.instantiateViewController(withIdentifier: "AboutNav")
        
        let communityVC = self.storyboard!.instantiateViewController(withIdentifier: "CommunityNav")
        
        explorerVC.tabBarItem = item(title: "My Programs", imageName: "programs")
        helpVC.tabBarItem = item(title: "Help", imageName: "help")
        aboutVC.tabBarItem = item(title: "About", imageName: "about")
        communityVC.tabBarItem = item(title: "Community", imageName: "community")
        
        self.viewControllers = [explorerVC, helpVC, aboutVC, communityVC]
                
        NotificationCenter.default.addObserver(self, selector: #selector(didAddProgram), name: NSNotification.Name(rawValue: "ProjectManagerDidAddProgram"), object: nil)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        AppController.shared.checkShowProgram()
    }
    
    override var keyCommands: [UIKeyCommand]? {
        if presentedViewController != nil {
            return nil
        }
        return [
            UIKeyCommand(input: "1", modifierFlags: .command, action: #selector(onTab1), discoverabilityTitle: "Show My Programs"),
            UIKeyCommand(input: "2", modifierFlags: .command, action: #selector(onTab2), discoverabilityTitle: "Show Help"),
            UIKeyCommand(input: "3", modifierFlags: .command, action: #selector(onTab3), discoverabilityTitle: "Show About"),
            UIKeyCommand(input: "4", modifierFlags: .command, action: #selector(onTab4), discoverabilityTitle: "Show Community")
        ]
    }
    
    func item(title: String, imageName: String) -> UITabBarItem {
        return UITabBarItem(title: title, image: UIImage(named: imageName), selectedImage: nil)
    }
    
    func dismissPresentedViewController(completion: @escaping () -> Void) {
        var topVC = self.selectedViewController
        if topVC is UINavigationController {
            topVC = (topVC as! UINavigationController).topViewController
        }
        if topVC?.presentedViewController != nil {
            topVC?.dismiss(animated: true, completion: completion)
        } else {
            completion()
        }
    }
    
    func showExplorer(animated: Bool, root: Bool) {
        self.selectedIndex = TabIndex.explorer.rawValue;
        let nav = self.selectedViewController as? UINavigationController
        if root {
            nav?.popToRootViewController(animated: animated)
        } else {
            /*
            if (![nav.topViewController isKindOfClass:[ExplorerViewController class]])
            {
                [nav popViewControllerAnimated:animated];
            }*/
        }
    }
    
    func showHelp(chapter: String) {
        self.selectedIndex = TabIndex.help.rawValue;
        let helpVC = self.selectedViewController as! HelpSplitViewController
        helpVC.showChapter(chapter)
    }
    
    @objc func didAddProgram() {
        if self.selectedIndex != 0 {
            self.selectedIndex = 0;
        }
    }
    
    @objc func onTab1() {
        selectedIndex = 0
    }
    
    @objc func onTab2() {
        selectedIndex = 1
    }
    
    @objc func onTab3() {
        selectedIndex = 2
    }
    
    @objc func onTab4() {
        selectedIndex = 3
    }
    
}
